### PR TITLE
Mixing entities and plain-text values

### PR DESCRIPTION
**Pull request by benbosman**
This PR is still work in progress, but has been created to allow for an early review.
The goal is to support using both entities and plain text for a metadata field and the approach is explained in https://docs.google.com/document/d/1wEmHirFzrY3qgGtRr2YBQwGOvH1IuTVGmxDIdnqvwxM/edit#heading=h.mw35favtpt8z

The order is visible in the virtual metadata, e.g.:
```json
{
  "metadata": [
    {
      "key": "dc.contributor.author",
      "value": "Simmons, Cameron",
      "language": "*",
      "place": 0,
      "authority": "virtual::732",
      "confidence": -1
    },
    {
      "key": "dc.contributor.author",
      "value": "Plain text author",
      "language": null,
      "place": 1,
      "authority": null,
      "confidence": -1
    },
    {
      "key": "dc.contributor.author",
      "value": "Plain text author 2",
      "language": null,
      "place": 2,
      "authority": null,
      "confidence": -1
    },
    {
      "key": "relation.isAuthorOfPublication",
      "value": "0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f",
      "language": "*",
      "place": 1,
      "authority": "virtual::889",
      "confidence": -1
    },
    {
      "key": "dc.contributor.author",
      "value": "Van de Velde, Kevin",
      "language": "*",
      "place": 4,
      "authority": "virtual::890",
      "confidence": -1
    }
  ]
}
```

The metadata values with an authority containing virtual and the ID of the relationship are the virtual metadata values.

Tasks to be completed are:
* Merge the latest configurable_entities branch into this branch
* JavaDoc, ITs, license headers
* There's still a small bug in calculating the places in some use cases

There's overlap with https://github.com/DSpace/DSpace/pull/2332 because CRUD on relationships helps to verify the functionality for determining the place when creating relationships.
The PR has already been created to facilitate an early review

Diff between this PR and #2332: https://github.com/atmire/DSpace/compare/w2p-57159_permission-to-create-relations...atmire:w2p-58898_place-column-calculation-error

### CODE REVIEW

**Comment by tdonohue** [comment]
@benbosman : Based on the description here, I'm seeing some possible parallels between this PR's effort and @cwilper's work on representing DSpaceObject metadata as a map...see:
* Ticket: https://jira.duraspace.org/browse/DS-4107
* REST Contract: https://github.com/DSpace/Rest7Contract/pull/39/
* REST Implementation: https://github.com/DSpace/DSpace/pull/2287

I wanted to point out these parallels since this PR is still a work-in-progress.  I wonder whether you feel @cwilper's work would help make this implementation *easier*?  If so, we can try and get Chris's work moved along more rapidly & note it as a dependency here.

**Comment by benbosman** [comment]
@tdonohue : We are aware of the overlap, which is why we kept that part very lightweight here.
Once Chris' work has been approved and merged, I would prefer to remove the changes here and use Chris' work instead.
The only change to revert in that case should be `MetadataEntryRest.java`.
I would prefer to avoid "waiting" for Chris' work to continue this work if we can continue (also in Angular) using just a few lines of code

**Comment by benbosman** [comment]
The latest updates to this PR include:
* Merged the latest configurable_entities branch into this branch
* Merged the latest https://github.com/DSpace/DSpace/pull/2332 code into this branch
* ITs (they are all succeeding), license headers
* Fixed the bug in calculating the places in the few broken use cases

JavaDoc still needs to be completed

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (AlexanderS) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: Missing JavaDoc for new public method.

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java
Comment (AlexanderS) [inline dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: So this is this?
```
relationship.setLeftPlace(leftRelationships.size());
```
I think this would be more readable then `leftRelationships.get(0).getLeftPlace() + 1` and it would even work for the case if the relationships are empty.

And is there any reason, why the sorting and reordering is done descending and from the back of the list to the front? Should not something like this also work?

```
leftRelationships.sort((o1, o2) -> o1.getLeftPlace() - o2.getLeftPlace());
for (int i = 0; i < leftRelationships.size(); i++) {
    leftRelationships.get(i).setLeftPlace(i);
}
```

File: dspace/config/spring/api/core-services.xml
Comment (AlexanderS) [inline dspace/config/spring/api/core-services.xml]: Is there any reason why a virtual metadata should not be *used for place*? Or is this just for performance if there are a lot of relations?

**Review by AlexanderS — COMMENTED** [review]
I just made a code review and added a few comments. Overall the code looks good.

It is a bit pity, that we need to duplicate some code because of the left and right side needs to be handled separately. But that cannot be changed.

File: dspace-api/src/main/java/org/dspace/content/DSpaceObjectServiceImpl.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/DSpaceObjectServiceImpl.java]: This has been updated to `metadataValues = getMetadata(dso, Item.ANY, Item.ANY, Item.ANY, Item.ANY);`
It removed the dependency on `ItemService`

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (benbosman) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: The JavaDoc has been written now

File: dspace/config/spring/api/core-services.xml
Comment (benbosman) [reply dspace/config/spring/api/core-services.xml]: This is both for performance and because a single relationship can map multiple metadata fields at once while calculation can only happen on one field (it is a place stored in the relationship)

If you have a relationship between an author and a publication, there's no need to start calculating the publication's place in the author using a metadata field in the author

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: @AlexanderS thanks for the great suggestion. It does indeed result in the same output but is more readable

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (AlexanderS) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: > Method to support updating a DSpace instance.

*DSpace instance*? Maybe *DSpaceObject instance*?

**Comment by benbosman** [comment]
All feedback has been processed now, and it's ready for the final review

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (benbosman) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: It's actually not limited to a DSpaceObject.
I've adjusted it to be similar to the patch comment:
`Apply an update to the REST object via JSON PUT`

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (AlexanderS) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: `DSpace instance` should be replace with `REST object` here, too.

**Review by AlexanderS — APPROVED** [review]
Looks good to me now. Just one little wording issue is left.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java
Comment (benbosman) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: Thanks, I pushed this wording change as well

**Comment by AlexanderS** [comment]
Note: This PR has merge conflicts, now.

**Comment by benbosman** [comment]
I was still running the Integration Tests locally, but the branch has now been updated with the latest configurable_entities branch

File: dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: Please consider that metadataValue can be null because constructMetadataValue might return null.

Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: Thanks @paulo-graca for identifying this. I've updated the branch with a check for null values from this method and some other small changes

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: Turn authorization back on

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: All of the hardcoded `text/uri-list` strings in your Tests should likely be replaced with `org.springframework.data.rest.webmvc.RestMediaTypes.TEXT_URI_LIST_VALUE`   We don't need to be hardcoding this over and over again, just like we don't hardcode `application/json` in our tests.

**Review by tdonohue — DISMISSED** [review]
@benbosman : I gave this PR a more thorough review today.  Overall, I think the code is looking good, but I think the tests could use some (minor) work/improvements. There's also some areas of code that I feel could be better commented / documented, as the code logic is not always self-explanatory or straightforward.

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: Why are we turning off authorization here?  I'm a little nervous anytime I see us turn off authorization *without* an inline comment to describe it.  Shouldn't the user have proper permissions to do these updates?

**Review by benbosman — COMMENTED** [review]
Thanks @tdonohue for your review.

Raf has committed improvements and I've answered your comments inline

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/RelationshipRestRepository.java
Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/RelationshipRestRepository.java]: @benbosman : I think we should add a comment to this `context.turnOffAuthorizationSystem();` call that says what you just stated. We should make it clear why we are turning off the authorization.

Comment (benbosman) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/RelationshipRestRepository.java]: We'll add a comment here

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This test method leaves the authorization system *off* for all tests.  Again, as I comment frequently, we need to ONLY turn off authorization *briefly* in tests. It is **not acceptable** to have tests which leave the authorization system turned off during the entirety of the test.

This method creates/updates  test content in several places...so it needs to be updated to surround all sections with `context.turnOffAuthorisationSystem();` and `context.restoreAuthSystemState();`

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: Just like above comment, this method is leaving authorization system turned off for the entire method. Please correct it.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: Again, authorization system is left turned off for the entire test method

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: And again, another new test method with authorization left turned off for the entire method.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (tdonohue) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: And, again, authorization left turned off here.

**Review by tdonohue — COMMENTED** [review]
@benbosman : I've reviewed this again today by doing a [diff between this PR and #2332](https://github.com/atmire/DSpace/compare/w2p-57159_permission-to-create-relations...atmire:w2p-58898_place-column-calculation-error).

The code itself looks good. But, there are still several (new) test methods that are incorrectly leaving all Authorization checks turned off.  We need to be very strict about not allowing this behavior (it's potentially dangerous -- as the tests may not always be fully trustworthy).  Once these tests are corrected, then I'll gladly add a +1 here.

File: dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: 👍

Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: @benbosman I noticed the added verifications, but the one I pointed out is still unverified. The `metadataValue.setAuthority` on line 1440.

Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: @paulo-graca if I'm not mistaken, you're referring to https://github.com/DSpace/DSpace/blob/970b66e/dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java#L1458 in the current codebase?
This is verifying whether metadata value is null.

Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java]: My fault! Thank you! 👍

**Review by paulo-graca — APPROVED** [review]
Thank you @benbosman !

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (benbosman) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This was resolved by https://github.com/DSpace/DSpace/pull/2331/commits/970b66edb2f9d8cd9a9824f0a1be266073c2a139#diff-61a3266941bcb44b91094e4d3f330ba6R790 and below

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (benbosman) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This was resolved by https://github.com/DSpace/DSpace/pull/2331/commits/970b66edb2f9d8cd9a9824f0a1be266073c2a139#diff-61a3266941bcb44b91094e4d3f330ba6R997 and below

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (benbosman) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This was resolved by https://github.com/DSpace/DSpace/pull/2331/commits/970b66edb2f9d8cd9a9824f0a1be266073c2a139#diff-61a3266941bcb44b91094e4d3f330ba6R1208 and below

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java
Comment (benbosman) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This was resolved by https://github.com/DSpace/DSpace/pull/2331/commits/970b66edb2f9d8cd9a9824f0a1be266073c2a139#diff-61a3266941bcb44b91094e4d3f330ba6R1331 and below

**Review by benbosman — COMMENTED** [review]
@tdonohue Raf had implemented these changes a few days ago, and has now also completed the merges to make sure this branch is compatible with configurable_entities.
It has been verified using both the ITs and manually using in the HAL browser

**Review by tdonohue — APPROVED** [review]
👍 This looks good to me now.  Thanks for the updates @benbosman !   As this has already received approval from @paulo-graca and @AlexanderS , I'll merge it shortly.

**Pull request merged**
